content/docs/testing-recipes.md

```diff
@@ -0,0 +1,403 @@
+
+```jsx
+// toggle.js
+import React, { useState } from "react";
+function Toggle(props) {
+  // initial, onChange
+  const [state, setState] = useState(Boolean(initial));
+  return (
+    <button
+      onClick={() => {
+        setState(previousState => !previousState);
+        onChange(!previousState);
+      }}
+      data-testid="toggle"
+    >
+      {state === true ? "Turn off" : "Turn on"}
+    </button>
+  );
+}
+```
+
+We could write tests for it:
+
+```jsx
+// toggle.test.js
+import React from "react";
+import { render } from "react-dom";
+import Toggle from "./toggle";
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+it("changes value when clicked", () => {
+  // a counter to track clicks
+  let clicks = 0;
+  act(() => {
+    render(<Toggle initial={true} onChange={() => clicks++} />, container);
+  });
+  // get a hold of the button element, and trigger some clicks on it
+  const button = document.querySelector("[data-testid=toggle]");
+  expect(button.innerText).toBe("Turn on!");
+  act(() => {
+    button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+  });
+  expect(clicks).toBe(1);
+  expect(button.innerText).toBe("Turn off!");
+  act(() => {
+    for (let i = 0; i < 5; i++) {
+      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+    }
+  });
+  expect(clicks).toBe(6);
+  expect(button.innerText).toBe("Turn on!");
+});
+```
+
+### Timers {#timers}
+
+UIs could use timer based functions like `setTimeout` to run time sensitive code. In this example, a multiple choice panel waits for a selection and advances, timing out if a selection isn't made in 5 seconds.
+
+```jsx
+function Card(props) {
+  useEffect(
+    props => {
+      const timeoutID = setTimeout(() => {
+        props.onSelect(null);
+      }, 5000);
+      return () => {
+        clearTimeout(timeoutID);
+      };
+    },
+    [props.onSelect]
+  );
+  return [1, 2, 3, 4].map(choice => {
+    <button
+      key={choice}
+      data-test-id={choice}
+      onClick={() => props.onSelect(choice)}
+    >
+      {choice}
+    </button>;
+  });
+}
+```
+
+We can write tests for this component by leveraging jest's timer mocks, and testing the different states it can be in.
+
+```jsx
+// card.test.js
+jest.useFakeTimers();
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+function render(ui){
+  act(() => {
+    ReactDOM.render(ui, container);
+  });
+}
+
+it("should select null after timing out", () => {
+  const yields = [];
```

Comment:
I think most people will be confused by the use of the word "yields" here

Reply:
Agreed. I think it'd be better to use a mock function instead.